core/src/execution/datafusion/expressions/bloom_filter_might_contain.rs

```diff
@@ -0,0 +1,165 @@
+        )
+    }
+}
+
+impl PartialEq<dyn Any> for BloomFilterMightContain {
+    fn eq(&self, _other: &dyn Any) -> bool {
+        down_cast_any_ref(_other)
+            .downcast_ref::<Self>()
+            .map(|other| {
+                self.bloom_filter_expr.eq(&other.bloom_filter_expr)
+                    && self.value_expr.eq(&other.value_expr)
+            })
+            .unwrap_or(false)
+    }
+}
+
+impl BloomFilterMightContain {
+    pub fn new(
+        bloom_filter_expr: Arc<dyn PhysicalExpr>,
+        value_expr: Arc<dyn PhysicalExpr>,
+    ) -> Self {
+        Self {
+            bloom_filter_expr,
+            value_expr,
+            bloom_filter: Default::default(),
+        }
+    }
+}
+
+impl PhysicalExpr for BloomFilterMightContain {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn data_type(&self, _input_schema: &arrow_schema::Schema) -> Result<DataType> {
+        Ok(DataType::Boolean)
+    }
+
+    fn nullable(&self, _input_schema: &arrow_schema::Schema) -> Result<bool> {
+        Ok(true)
+    }
+
+    fn evaluate(&self, batch: &RecordBatch) -> DataFusionResult<ColumnarValue> {
+        // lazily get the spark bloom filter
+        if self.bloom_filter.get().is_none() {
+            let bloom_filter_bytes = self.bloom_filter_expr.evaluate(batch)?;
+            match bloom_filter_bytes {
+                ColumnarValue::Array(_) => {
+                    return internal_err!(
+                        "Bloom filter expression must be evaluated as a scalar value"
+                    );
+                }
+                ColumnarValue::Scalar(ScalarValue::Binary(v)) => {
+                    let filter = v.map(|v| SparkBloomFilter::new_from_buf(v.as_bytes()));
+                    self.bloom_filter.get_or_init(|| filter);
+                }
+                _ => {
+                    return internal_err!("Bloom filter expression must be binary type");
+                }
+            }
+        }
+        let num_rows = batch.num_rows();
+        let lazy_filter = self.bloom_filter.get().unwrap();
+        if lazy_filter.is_none() {
+            // when the bloom filter is null, we should return a boolean array with all nulls
+            Ok(ColumnarValue::Array(Arc::new(BooleanArray::new_null(
+                num_rows,
+            ))))
+        } else {
+            let spark_filter = lazy_filter.as_ref().unwrap();
+            let values = self.value_expr.evaluate(batch)?;
+            match values {
+                ColumnarValue::Array(array) => {
+                    let array = array
+                        .as_any()
+                        .downcast_ref::<Int64Array>()
+                        .expect("value_expr must be evaluated as an int64 array");
+                    Ok(ColumnarValue::Array(Arc::new(
+                        spark_filter.might_contain_longs(array)?,
+                    )))
+                }
+                ColumnarValue::Scalar(a) => match a {
```

Comment:
nit: maybe `ColumnarValue::Scalar(ScalarValue::Int64(v)) => `?

Reply:
And the `internal_err!` can be pulled up. It should be along with array and scalar cases.